client/src/main/java/io/pravega/client/stream/SegmentReader.java

```diff
@@ -0,0 +1,56 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.pravega.client.stream;
+
+/**
+ * Allows to read data from a specific segment.
+ */
+public interface SegmentReader<T> extends AutoCloseable {
+
+    /**
+     * Gets the next event in the segment. If there are no events currently available this will block up for
+     * timeoutMillis waiting for them to arrive. If none do, an EventReadWithStatus {@link EventReadWithStatus} will be returned with null for
+     * {@link EventReadWithStatus#getEvent()}.
+     *
+     * @param timeoutMillis An upper bound on how long the call may block before returning null.
+     * @return An instance of {@link EventReadWithStatus}, which contains the next event in the segment. In the case the timeoutMillis
+     *         is reached, {@link EventReadWithStatus#getEvent()} returns null.
+     */
+    EventReadWithStatus<T> read(long timeoutMillis);
```

Comment:
Take a look at how https://cwiki.apache.org/confluence/display/FLINK/FLIP-27%3A+Refactor+Source+Interface#FLIP27:RefactorSourceInterface-SourceReaderSourceReader  `pollNext` and `isAvailable` are implemented. The read method itself is non-blocking but sync, where as isAvailable returns a future. 

We can implement this by having the impl use a SegmentInputStream's read with a buffer no larger than the the value returned via `bytesInBuffer` and use it's `fillBuffer` method to request more data.